File: src/ru/hudspeed/pro/MapKitLifecycle.java
package ru.gpsantiradar.app;

public final class MapKitLifecycle {
    public interface Delegate {
        void onStart();
        void onStop();
    }

    private final Delegate delegate;
    private int owners;

    public MapKitLifecycle(Delegate delegate) {
        this.delegate = delegate;
    }

    public synchronized void acquire() {
        if (owners++ == 0) delegate.onStart();
    }

    public synchronized void release() {
        if (owners == 0) return;
        if (--owners == 0) delegate.onStop();
    }
}

File: src/main/kotlin/ru/gpsantiradar/app/MapKitLifecycle.kt
package ru.gpsantiradar.app

class MapKitLifecycle(private val delegate: Delegate) {
    interface Delegate { fun onStart(); fun onStop() }
    private var owners = 0

    @Synchronized fun acquire() { if (owners++ == 0) delegate.onStart() }
    @Synchronized fun release() {
        if (owners == 0) return
        if (--owners == 0) delegate.onStop()
    }
}
